xds/src/main/java/io/grpc/xds/LookasideLb.java

```diff
@@ -112,4 +122,38 @@ public LoadBalancer newLoadBalancer(Helper helper) {
   interface LookasideChannelLbFactory {
     LoadBalancer newLoadBalancer(Helper helper, AdsStreamCallback adsCallback, String balancerName);
   }
+
+  private static final class LookasideChannelLbFactoryImpl implements LookasideChannelLbFactory {
```

Comment:
You do not need both `LoadBalancerProvider` and `LookasideChannelLbFactory`. You `LookasideLb`'s secondary constructor can just take a `LoadBalancerProvider` that creates `LookasideChannelLb` instances. The implementation for how to create `LookasideChannelLb` instances can be private. In your test, you can fake the `LoadBalancerProvider` to capture the `helper`.

```java
LookasideLb(Helper helper, ...) {
   this(..., new LookasideChannelLbProvider(...));
}

LookasideLb(..., LoadBalancerProvider provider) {
   ...
}

private class LookasideChannelLbProvider extends LoadBalancerProvider {
   ...   

   @Override
   public LoadBalancer newLoadBalancer(Helper helper) {
      // Create a LookasideChannelLb instance.
   }
}

```

Reply:
Then this provider will have a `balancerName` that can be changed. It won't work for graceful switch. Even it can work after some hack, I'm scared of having the `LookasideChannelLbProvider` depending on the state of `LookasideLb`. It should be totally independent of and have no knowledge of the state of `LookasideLb`.